README.md
# simpletron
The Simpletron, software-based simulation of a primitive computer. 

## Description
The simpletron equipped with 100 words of memory. Words can be referenced by their location number 00 .. 99.
Word is a signed 4 digit decimal as +0001, +3364, -1293.
Word can be data or instruction. Instruction always positive; data can be positive or negative. First, two digits of the word are operation second operands. 

Simpletron CPU contains an accumulator and can handle basic arithmetic operations, memory and console IO operations, basic program branching. 


## App Setup

### Prerequisites

* Installed [Docker](https://www.docker.com/)

or for local setup:
* Installed [clang](https://clang.llvm.org/)
* Installed [gnu make](https://www.gnu.org/software/make/)
* Installed [ctags](https://github.com/universal-ctags/ctags)

### Build and run Docker

```bash
git clone git://github.com/cpputest/cpputest.git
docker-compose up -d
./docker-make-tests.sh
./docker-make.sh
./docker-simpletron.sh
```

Get interactive shell in docker container 
```bash
./docker-shell.sh
```

### Build and run local 

```bash
git clone git://github.com/cpputest/cpputest.git
./make-tests.sh
make simpletron
bin/simpletron
```

## Operation manual

### Command and data format
DATA - four digit interger (-9999 .. +9999)

COMMAND - four digit positive integer, first two digit operation code last two digit operand 
```
1010 - read data from counsole to memory address 0010
4003 - go to 3rd instruction
```

After the first prompt, you need to enter the file name, which contains a file with machine codes without an empty line at the end.
Program text example:
```
1007
1008
2007
3008
2109
1109
4300
```
If simpletron loaded the program successfully, then the execution begins.

The program can ask for user input.
Input must be data word  (-9999 .. +9999)

All console output repeated to simpletron_out.txt file. 
### Supported commands

* READ  
  ```CPP
	#define READ 10	// Read a word from the terminal into 
				//a specific location in memory
  ```
   format +1050 - 10 operation code 50 operand - memory address 

* WRITE
  ```CPP
	#define WRITE 11	//Write a word from specific location
						//in memory to the terminal
  ```
  format +1150 - 11 operation code 50 operand - memory address 

* LOAD
  ```CPP
	#define LOAD 20		// Load a word from a specific memory location 
		   				// into accumulator
  ```

* STORE
  ```CPP
	#define STORE 21	// Store a word from accumulator into
						// a specific location in memory
   ```
* ADD
  ```CPP
	#define ADD 30		// Add word from specific memory location 
						// to accumulator
  ```

* SUBSTRACT
  ```CPP
	#define SUBSTRACT 31	// Substract  word from specific memory location 
							// from accumulator
  ```
* DIVIDE
  ```CPP
	#define DIVIDE 32		// Devide a word from accumulator 
							// by word from specific memory location 
  ```
* MULTIPLY
  ```CPP
	#define MULTIPLY 33		// Multiply a word from accumulator 
							// by word from specific memory location 
  ```
* BRANCH
  ```CPP
	#define BRANCH 40		// Branch to a specific location in memory
  ```
* BRANCHNEG
  ```CPP
	#define BRANCHNEG 41	// Branch to a specific location in memory if 
							// the accumulator is negative
  ```
* BRANCHZERO
  ```CPP
	#define BRANCHZERO 42	// Branch to a specific location in memory if 
							// the accumulator is zero
  ```

### Program examples 

#### Sum of two variables

```
1007 // Read A and save to address 0007
1008 // Read B ans save to address 0008
2007 // Load A to accumulator
3008 // Add B
2109 // store accumulator to C (address 0009)
1109 // output C value to console
4300 // end of program
```
Output:
```
*** Program execution begins. ***
 ? 7
 ? 3
 > +0010
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			+0010
instrcutionCounter:		   06
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1007	+1008	+2007	+3008	+2109	+1109	+4300	+0007	+0003	+0010
10	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
```

#### Get a larger value of two

```
1009 // Read A to address 0009
1010 // Read B to address 0010
2009 // Load to accumulator
3110 // Substract B from accumulator
4107 // if accumulator negative go to 07
1109 // Write A to console
4300 // End of program
1110 // Write B to console 
4300 // End of program
```

Output:
```
*** Program execution begins. ***
 ? 5
 ? 10
 > +0010
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			-0005
instrcutionCounter:		   08
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1009	+1010	+2009	+3110	+4107	+1109	+4300	+1110	+4300	+0005
10	+0010	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
```

```
*** Program execution begins. ***
 ? 7
 ? 3
 > +0007
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			+0004
instrcutionCounter:		   06
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1009	+1010	+2009	+3110	+4107	+1109	+4300	+1110	+4300	+0007
10	+0003	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
```

#### Get a sum of positive numbers
calculate the sum of positive numbers and exit if entered a negative number

```
1099 // Input A (address 0099)
2099 // Load A to Accumulator
4106 // if Accumulator < 0 got to 0006
3098 // Accumulator += SUM (address 0098)
2198 // SUM = Accumulator
4000 // go to 0000
1198 // Print SUM 
4300 // End 
```
 Output:
 ```
 *** Program execution begins. ***
 ? 1
 ? 1
 ? 1
 ? 10
 ? 20
 ? -1
 > +0033
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			-0001
instrcutionCounter:		   07
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1099	+2099	+4106	+3098	+2198	+4000	+1198	+4300	+0000	+0000
10	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0033	-0001
 ```

#### Calculate average
```
1099 // Get Number of arguments 
1098 // Step for counter
2099 // Accumulator = Number of arguments
2197 // Counter = Accumulator
1096 // Get Nth argument
2096 // Accumulator = Argument 
3095 // Accumulator += Sum
2195 // Sum = Accumulator
2097 // Accumulator = Counter
3198 // Accumulator -= Step
4213 // If Accumulator == 0 go to 13
2197 // Counter = Accumulator 
4004 // Go to 04
2095 // Accumulator = Sum 
3299 // Accumulator /= Number of Arguments
2194 // Result = Accumulator
1194 // Print result 
4300 // End
```
Output:
```
*** Program execution begins. ***
 ? 5 
 ? 1 
 ? 1 
 ? 2
 ? 3
 ? 4
 ? 5
 > +0003
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			+0003
instrcutionCounter:		   17
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1099	+1098	+2099	+2197	+1096	+2096	+3095	+2195	+2097	+3198
10	+4213	+2197	+4004	+2095	+3299	+2194	+1194	+4300	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0003	+0015	+0005	+0001	+0001	+0005
```

#### Get max number from numbers series 
```
1099 // Read default max number 
1098 // Read step of iteration
1097 // Read Series size (Counter)
1096 // Read Nth series number
2096 // Accumulator = Nth number
3199 // Accumulator -= max number
4109 // If Accumulator < 0 go to 09
2096 // Accumulator = Nth number
2199 // Max number = Accumulator
2097 // Accumulator = Iteration counter
3198 // Accumulator -= iteration step
2197 // Couunter = Accumulator
4214 // If counter == 0 go to 14
4003 // go to 3
1199 // Print Max number
4300 // End
```
Output:
```
*** Program execution begins. ***
 ? -1000
 ? 1
 ? 5
 ? 100
 ? -200
 ? 5
 ? 3
 ? 3
 > +0100
*** Simpletron execution terminated ***
REGISTERS:
accumulator:			+0000
instrcutionCounter:		   15
instructionRegister:		+4300
operationCode:			   43
operand:			   00

MEMORY:
 	    0	    1	    2	    3	    4	    5	    6	    7	    8	    9
0	+1099	+1098	+1097	+1096	+2096	+3199	+4109	+2096	+2199	+2097
10	+3198	+2197	+4214	+4003	+1199	+4300	+0000	+0000	+0000	+0000
20	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
30	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
40	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
50	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
60	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
70	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
80	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000	+0000
90	+0000	+0000	+0000	+0000	+0000	+0000	+0003	+0000	+0001	+0100
```


CHANGELOG.md
## [v0.4.0] - 2022-01-12

### Added 
    - Possibility to use docker as build and run environment
    - unit test coverage by CPPUTEST

### Changed
    - modules structure

## [v0.3.0] - 2021-12-20

### Changed 
    - project structure
    - Makefile
    - README.md

## [v0.2.0] - 2021-07-27

### Added
    - program reading from file
    - repeating program output to file

### Deleted
    - possibility to read the program from console

## [v0.1.1] - 2021-07-26

### Added
    - Check for illegal operation in runtime

## [v0.1.0] - 2021-07-22

### Added
    - Basic CPU, MEMORY, IO implementation
    - Support for program input from the console
    - Support for interaction with the program using a console


src/modules/CPU.c
#include <stdio.h>
#include <stdbool.h>

#include "CPU.h"
#include "RAM.h"
#include "IO.h"
#include "IO_showDataWord.h"
#include "IO_showAccumulatorOverflowMessage.h"
#include "IO_showDivideByZeroMessage.h"

//CPU commands
#define HALT 43

#define READ 10 		// Read a word from the terminal into 
						// a specific location in memory

#define WRITE 11		// Write a word from specific location
						//in memory to the terminal

#define LOAD 20			// Load a word from a specific memory location 
		   				// into accumulator

#define STORE 21		// Store a word from accumulator into
						// a specific location in memory
					
#define ADD 30 			// Add word from specific memory location 
						// to accumulator

#define SUBSTRACT 31	// Substract  word from specific memory location 
						// from accumulator

#define DIVIDE 32		// Devide a word from accumulator 
						// by word from specific memory location 

#define MULTIPLY 33		// Multiply a word from accumulator 
						// by word from specific memory location 

#define BRANCH 40		// Branch to a specific location in memory

#define BRANCHNEG 41	// Branch to a specific location in memory if 
						// the accumulator is negative

#define BRANCHZERO 42	// Branch to a specific location in memory if 
						// the accumulator is zero

static int accumulator = 0;
static size_t instructionCounter = 0;
static int instructionRegister = 0;
static int operationCode = 0;
static int operand = 0;
static bool overflowFlag = false;

static void read(int * memory, size_t address, FILE *outputFile);
static void write(int * memory, size_t address, FILE *outputFIle);
static void load(int * memory, size_t address);
static void store(int * memory, size_t address);
static void add(int * memory, size_t address);
static void substract(int *memory, size_t address);
static void divide(int *memory, size_t address, FILE *outputFile);
static void multiply(int *memory, size_t address);
static void branch();
static void branchneg();
static void branchzero();
static bool isAccumulatorOverflow();
static void checkAccumulatorOverflow();
static void goToValidLocation();

void CPU_executeProgram(int *memory, FILE *outputFile)
{
	accumulator = 0;
	instructionCounter = 0;
	instructionRegister = 0;
	operationCode = 0;
	operand = 0;
	overflowFlag = false;

	IO_showExecutionBeginsMessage(outputFile);

	while (instructionCounter < MEMORY_SIZE  && operationCode != HALT && !overflowFlag)
	{	
		instructionRegister = RAM_read(memory, instructionCounter);
		operationCode = instructionRegister / 100;
		operand = instructionRegister % 100;
		
		switch (operationCode)
		{
			case READ: 
				read(memory, operand, outputFile);
				break;
			case WRITE :
				write(memory, operand, outputFile);
				break;
			case LOAD:
				load(memory, operand);
				break;
			case STORE:
				store(memory, operand);
				break;
			case ADD:
				add(memory, operand);
				break;
			case SUBSTRACT:
				substract(memory, operand);
				break;
			case DIVIDE:
				divide(memory, operand, outputFile);
				break;
			case MULTIPLY:
				multiply(memory, operand);
				break;
			case BRANCH:
				branch();
				continue;
			case BRANCHNEG:
				branchneg();
				continue;
			case BRANCHZERO:
				branchzero();
				continue;
			case HALT:
				break;
			default:
			  IO_showMessageInvalidCommand(operationCode, (int) instructionCounter, outputFile);
			  overflowFlag = true;
		} //end switch operation code

		instructionCounter++;

	} // end while main loop

	instructionCounter--;
	IO_showExecutionTerminatedMessage(outputFile);

} // end function CPU_executeProgram

int CPU_getAccumulator()
{
	return accumulator;
} // end function CPU_getAccumulator

size_t CPU_getInstructionCounter()
{
	return instructionCounter;
} // end function CPU_getInstructionCounter

int CPU_getInstructionRegister()
{
	return instructionRegister;
} // end function CPU_getInstructionRegister

int CPU_getOperand()
{
	return operand;
} // end function CPU_getOperand

int CPU_getOperationCode()
{
	return operationCode;
} // end function CPU_getOperationCode

static void read(int * memory, size_t address, FILE *outputFile)
{
	IO_showInputPrompt();
	RAM_write(memory, address, IO_getValidDataWord(outputFile));
} // end function read

static void write(int * memory, size_t address, FILE *outputFile)
{
	IO_showDataWord(RAM_read(memory, address),outputFile);	
} // end function write

static void load(int * memory, size_t address)
{
	accumulator = RAM_read(memory, address);
} // end function load

static void store(int * memory, size_t address)
{
	RAM_write(memory, address, accumulator);
} // end function store

static void add(int * memory, size_t address)
{
	accumulator += RAM_read(memory, address);
	checkAccumulatorOverflow();
} // end function add 

static void substract(int * memory, size_t address)
{
	accumulator -= RAM_read(memory, address);
	checkAccumulatorOverflow();
} // end function substract 

static void divide(int * memory, size_t address, FILE *outputFile)
{
	int divider = RAM_read(memory, address);
	if (divider == 0)
	{
		IO_showDivideByZeroMessage(outputFile);
		overflowFlag = true;
		return;
	}
	accumulator /= divider; 
	checkAccumulatorOverflow();
} // end function divide 

static void multiply(int * memory, size_t address)
{
	accumulator *= RAM_read(memory, address);
	checkAccumulatorOverflow();
} // end function multiply 

static void branch()
{
	goToValidLocation();
} // end function branch

static void branchneg()
{
	if (accumulator < 0)
	{
		goToValidLocation();	
	} // end if check accumulator is negative 
	else 
	{
		instructionCounter++;
	} // end else if check accumulator negative
} // end function branchneg

static void branchzero()
{
	if (accumulator == 0)
	{
		goToValidLocation();
	} // end if check accumulator is zero
	else 
	{
		instructionCounter++;
	} // end else if check accumulator zero
} // end function branczero

static bool isAccumulatorOverflow()
{
	return (accumulator < LOW_DATA_LIMIT || accumulator > HIGH_DATA_LIMIT);
} // end function isAccumulatorOverflow

static void checkAccumulatorOverflow(FILE *outputFile)
{
	if (isAccumulatorOverflow())
	{
		IO_showAccumulatorOverflowMessage(outputFile);
		overflowFlag = true;		
	} // end if checkOverflow
} // end function checkAccumulatorOverflow

static void goToValidLocation(FILE *outputFile)
{
	if (operand < MEMORY_SIZE)
	{
		instructionCounter = operand;
	}
	else
	{
		IO_showOutOfMemoryMessage(outputFile);
		overflowFlag = true;
	} // end if else check memory limit
} // end function goToLocationWithValidation
